### ISSUE TITLE
Recursion causes segfault

### ISSUE THREAD

**Issue by Jackojc** [wotpp/wotpp#4]
```
let a a
a
```
Recursion in wot++ doesn't really make sense currently as there is no way to break out of a function through conditionals etc.

We have 2 options here:
1. Set a max recursion depth
2. Disallow recursive calls

**Comment by Jackojc** [comment]
As per @noxabellus' suggestion on Discord, we could also just look into supporting conditionals.

**Comment by UndefinedBHVR** [comment]
Could add an `escape` or `break` call in order to immediately escape recursion?

**Comment by qookei** [comment]
The problem is that even if we have `escape` or whatever, we still have no way to conditionally break out of recursion (due to no conditional support).

**Comment by Jackojc** [comment]
Just to note, conditionals are in the works.

**Comment by Jackojc** [comment]
Conditionals are now implemented so this issue has become relevant again.

**Comment by jlagarespo** [comment]
Since now we have conditionals, I guess we should implement a max recursion limit. Also, I see that arg parsing is in the works, perhaps we could have a command line option for that.

**Comment by Jackojc** [comment]
Had a talk about it the other day but I think the conclusion was that it doesn't really make sense to impose a recursion limit for a number of reasons:

1. You may actually have a valid reason to have a very deep recursive call
2. Where do you even set the recursion limit? The user can still use `ulimit` to shrink the stack to the point where it causes a stack overflow anyway
3. There isn't really a good way to check for stack overflows before they happen and recover

A middle ground we could try and compromise on is to detect very simple infinite recursion cases like:
```
let func
    func
func
```

but we have this thing called the halting problem that probably wont let us get too far with that lol

**Comment by jlagarespo** [comment]
I agree, but regardless of the halting problem, I don't think it would be too hard to check for a function that unconditionally calls itself and throw a warning.

**Comment by Jackojc** [comment]
We could give it a try.

**Comment by jlagarespo** [comment]
If we find more complex cases such as something like
```wpp
let a b
let b a
a
```
we might as well just give up, but for a function that calls itself outside of a map, we could display a nice little warning message.

**Comment by Jackojc** [comment]
Yeah sounds good

**Comment by Jackojc** [comment]
Maybe add a user configurable maximum stack depth through a commandline flag?

**Comment by Jackojc** [comment]
Added a warning for deep recursion on the dev branch. Not a solution to stack overflow but it helps track down recursion without an exit condition.
